optuna/integration/mlflow.py

```diff
@@ -0,0 +1,141 @@
+
+            mlflc = MLflowCallback(
+                tracking_uri=YOUR_TRACKING_URI,
+                metric_name='my metric score',
+            )
+
+            study = optuna.create_study(study_name='my_study')
+            study.optimize(objective, n_trials=10, callbacks=[mlflc])
+
+        .. testcleanup::
+
+            import shutil
+            shutil.rmtree(tempdir)
+
+        .. testoutput::
+            :hide:
+            :options: +NORMALIZE_WHITESPACE
+
+            INFO: 'my_study' does not exist. Creating a new experiment
+
+    Args:
+        tracking_uri:
+            Set the tracking server URI.
+
+            - An empty string, or a local directory path, prefixed with ``file:``. Data is stored
+              locally in the provided directory (or ``./mlruns`` if empty).
+            - An HTTP URI like ``https://my-tracking-server:5000``.
+            - A Databricks workspace, provided as the string ``databricks`` or, to use a
+              Databricks CLI.
+              `profile <https://github.com/databricks/databricks-cli#installation>`_,
+              ``databricks://<profileName>``.
+        experiment:
+            Name of MLflow experiment to be activated. If not set ``study.study_name``
+            will be taken. Either ``experiment`` or ``study.study_name`` must be set.
+        metric_name:
+            Name of the metric. If not provided this will be called ``trial_value``.
+    """
+
+    def __init__(self, tracking_uri=None, experiment=None, metric_name=None):
+        # type: (str, str, str) -> None
+
+        _check_mlflow_availability()
+
+        self._tracking_uri = tracking_uri
+        self._experiment = experiment
+        self._metric_name = metric_name
+
+    def __call__(self, study, trial):
+        # type: (optuna.study.Study, optuna.structs.FrozenTrial) -> None
+
+        # This sets the tracking_uri for MLflow.
+        if self._tracking_uri is not None:
+            mlflow.set_tracking_uri(self._tracking_uri)
+
+        # This sets the experiment of MLflow.
+        if self._experiment is not None:
+            mlflow.set_experiment(self._experiment)
+        elif (
+            study.study_name is not None
+            and study.study_name != "no-name-00000000-0000-0000-0000-000000000000"
+        ):
+            mlflow.set_experiment(study.study_name)
+        else:
+            raise ValueError("Either 'experiment' or 'study.study_name' must be set!")
+
+        with mlflow.start_run(run_name=trial.number):
+
+            # This sets the metric for MLflow.
+            trial_value = trial.value if trial.value is not None else float("nan")
+            metric_name = self._metric_name if self._metric_name is not None else "trial_value"
+            mlflow.log_metric(metric_name, trial_value)
+
+            # This sets the params for MLflow.
+            mlflow.log_params(trial.params)
+
+            # This sets the tags for MLflow.
+            tags = {}  # type: Dict[str, str]
```

Comment:
Those pieces of information are similar to what we can get with [Study.trials_dataframe](https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.Study.trials_dataframe). We may want to be consistent about the naming convention of their keys.

How about renaming keys as:
- `number` instead of `trial_number`
- `datetime_start` instead of `trial_datetime_start`
- `datetime_complete` instead of `trial_datetime_complete`
- `state` instead of `trial_state`

Reply:
I think naming the metric would be a very useful feature for me. I have cases where I do different experiments and change the metric. From F1 to average precision or from accuracy to ROC_AUC. Giving the metric an optional explicit name is very useful in these cases.

Reply:
> Those pieces of information are similar to what we can get with [Study.trials_dataframe](https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.Study.trials_dataframe). We may want to be consistent about the naming convention of their keys.
> 
> How about renaming keys as:
> 
>     * `number` instead of `trial_number`
> 
>     * `datetime_start` instead of `trial_datetime_start`
> 
>     * `datetime_complete` instead of `trial_datetime_complete`
> 
>     * `state` instead of `trial_state`

Removed the prefixes.

Reply:
> Removed the prefixes.

Thanks for your quick change!

> I think naming the metric would be a very useful feature for me.

This feature might be worth, but I think we should at first take the simpler choice of omitting the parameter given the maintenance cost. When a different metric name is necessary, the user has an alternative way of adding another metric outside this callback.

Reply:
Well, honestly. For me it is very importat to be able to change the metric name to something else then just "value". Here is a suggestion that reduces comeplexity (maintainance cost) but still makes it possible to change the metric name after object creation.

``` python
    def __init__(self, tracking_uri=None):
        # type: (Optional[str]) -> None

        _check_mlflow_availability()

        self._tracking_uri = tracking_uri
        self._METRIC_NAME = "value"

    def __call__(self, study, trial):
        # type: (optuna.study.Study, optuna.structs.FrozenTrial) -> None

        # This sets the tracking_uri for MLflow.
        if self._tracking_uri is not None:
            mlflow.set_tracking_uri(self._tracking_uri)

        # This sets the experiment of MLflow.
        if (
            study.study_name is not None
            and study.study_name != "no-name-00000000-0000-0000-0000-000000000000"
        ):
            mlflow.set_experiment(study.study_name)
        else:
            raise ValueError(
                "You must set 'study.study_name' inside Optuna! This is needed to set the MLflow experiment name."
            )

        with mlflow.start_run(run_name=str(trial.number)):

            # This sets the metric for MLflow.
            trial_value = trial.value if trial.value is not None else float("nan")
            metric_name = self._metric_name if self._metric_name is not None else "value"
            mlflow.log_metric(self._METRIC_NAME, trial_value)

```

What do you think?

Reply:
I don't think a private attribute is a good compromise as it clearly communicates to the user that it shouldn't be accessed/modified. If this feature is requested by MLflow users, I believe that we could provide it with a proper API, as in the current implementation.

Reply:
Ok, thanks. So the implementation stays as it is.